packages/network-controller/src/rpc-service/rpc-service.ts

```diff
@@ -0,0 +1,384 @@
+ * after request parameters have been validated.
+ *
+ * This is somewhat difficult to verify because JavaScript engines (and in
+ * some cases libraries) produce slightly different error messages for this
+ * particular scenario, and we need to account for this.
+ *
+ * @param error - The error.
+ * @returns True if the error indicates that the network is unreachable, and
+ * false otherwise.
+ */
+export default function isNetworkUnreachableError(error: unknown) {
+  return (
+    error instanceof TypeError && NETWORK_UNREACHABLE_ERRORS.has(error.message)
+  );
+}
+
+/**
+ * Guarantees a URL, even given a string. This is useful for checking components
+ * of that URL.
+ *
+ * @param endpointUrlOrUrlString - Either a URL object or a string that
+ * represents the URL of an endpoint.
+ * @returns A URL object.
+ */
+function getNormalizedEndpointUrl(endpointUrlOrUrlString: URL | string): URL {
+  return endpointUrlOrUrlString instanceof URL
+    ? endpointUrlOrUrlString
+    : new URL(endpointUrlOrUrlString);
+}
+
+/**
+ * This class is responsible for making a request to an endpoint that implements
+ * the JSON-RPC protocol. It is designed to gracefully handle network and server
+ * failures, retrying requests using exponential backoff. It also offers a hook
+ * which can used to respond to slow requests.
+ */
+export class RpcService implements AbstractRpcService {
+  /**
+   * The function used to make an HTTP request.
+   */
+  readonly #fetch: typeof fetch;
+
+  /**
+   * The URL of the RPC endpoint.
+   */
+  readonly #endpointUrl: URL;
+
+  /**
+   * A common set of options that the request options will extend.
+   */
+  readonly #fetchOptions: FetchOptions;
+
+  /**
+   * The policy that wraps the request.
+   */
+  readonly #policy: ServicePolicy;
+
+  /**
+   * Constructs a new RpcService object.
+   *
+   * @param args - The arguments.
+   * @param args.fetch - A function that can be used to make an HTTP request.
+   * If your JavaScript environment supports `fetch` natively, you'll probably
+   * want to pass that; otherwise you can pass an equivalent (such as `fetch`
+   * via `node-fetch`).
+   * @param args.btoa - A function that can be used to encode a binary string
+   * into base 64. Used to encode authorization credentials.
+   * @param args.endpointUrl - The URL of the RPC endpoint.
+   * @param args.fetchOptions - A common set of options that will be used to
+   * make every request. Can be overridden on the request level (e.g. to add
+   * headers).
+   * @param args.failoverService - An RPC service that represents a failover
+   * endpoint which will be invoked while the circuit for _this_ service is
+   * open.
+   */
+  constructor({
+    fetch: givenFetch,
+    btoa: givenBtoa,
+    endpointUrl,
+    fetchOptions = {},
+  }: {
+    fetch: typeof fetch;
+    btoa: typeof btoa;
+    endpointUrl: URL | string;
+    fetchOptions?: FetchOptions;
+    failoverService?: AbstractRpcService;
```

Comment:
Nit: Was this left by mistake?
```suggestion
    failoverService?: AbstractRpcService;
```

Reply:
Oops, thank you! Yes, this is for another PR.